### ISSUE TITLE
Linux dedicated server not loading mods

### ISSUE THREAD

**Issue by arousle** [Regalis11/Barotrauma#11855]
### Disclaimers

- [X] I have searched the issue tracker to check if the issue has already been reported.
- [X] My issue happened while using mods. (sort of)

### What happened?

I'm running a dedicated Barotrauma server (using the [yanwk/barotrauma-server Docker image](https://github.com/YanWenKun/barotrauma-server-docker), not that I think it matters). The server works fine for running a Vanilla campaign, but I want to have some mods running on the server, too, and I've spent half the day unsuccessfully trying to make that happen.

As I understand it, you need to do two things for a mod to be loaded by the dedicated server executable:

1. Put the mod files in a local directory on the server.
2. Specify the mod as a content package in `config_player.xml`, pointing it to the directory from 1).

My minimal `config_player.xml` for loading a single mod looks like this:

```xml
<config
  language="English">
  <contentpackages>
    <!--Vanilla-->
    <corepackage
      path="Content/ContentPackages/Vanilla.xml" />
    <regularpackages>
      <!--Some mod-->
      <package path="Daedalic Entertainment GmbH/Barotrauma/WorkshopMods/Installed/2962604067/filelist.xml" />
    </regularpackages>
  </contentpackages>
</config>
```

However, with this setup, the server starts up without errors, but doesn't actually seem to load any mods. I don't see the mod loaded in the game's server browser information window, and if I actually start round I don't see the mod's effects in the game, either. I suspected the `package path` as the culprit, but I've tried every permutation of absolute/relative paths I could think of, all with the same result. 

At this point I've gone through through so many different permutations of `config_player.xml` that I'm no longer sure it's an issue on my end, so I'm opening this issue here. `config_player.xml` _does_ get loaded by the server, if I change e.g. the language to Chinese I can see the change being reflected in the server logs.

### Reproduction steps

1. Host a dedicated Barotrauma server on Linux (in my case, on Ubuntu 22.04 running Docker 20.10.21 and the [yanwk/barotrauma-server Docker image](https://github.com/YanWenKun/barotrauma-server-docker)).
2. Place one or more mods in the `Daedalic Entertainment GmbH/Barotrauma/WorkshopMods/Installed/` directory (or, I assume, wherever) and include them in the `config_player.xml` of the server as content packages as described above.
3. Find your server in the Barotrauma game client server list. It will show that no mods besides Vanilla are enabled.

### Bug prevalence

Happens every time I play

### Version

v1.0.9.0

### Which operating system did you encounter this bug on?

Linux

### Relevant error messages and crash reports

```shell
None. The server starts up fine, but only has the Vanilla content package enabled.
```

**Comment by Regalis11** [comment]
Thank you for the report!

The paths in the config files should be absolute (unless the file is inside the game's installation folder, like "Content/ContentPackages/Vanilla.xml", in which case you can use a path relative to the installation folder). You did mention trying absolute paths too, so I guess the issue must be somewhere else if those paths were correct.

We have [a page on the wiki describing how to enable mods on the dedicated server](https://barotraumagame.com/wiki/Enabling_Mods_on_a_Dedicated_Server) as well, please take a look in case you haven't already!  It seems that there are also some extra things that need to be taken into consideration when using [the Docker image](https://github.com/YanWenKun/barotrauma-server-docker), but I unfortunately don't have any experience in using Docker images so I cannot provide any additional help on that.

We also [got another report](https://github.com/Regalis11/Barotrauma/issues/11699) recently about similar difficulties with enabling mods on a Linux dedicated server, although I'm not sure if that's relevant here.

I'm afraid I can't say for sure where the issue might be off the top of my head, but I hope some of those links can give you some more clues on diagnosing where the problem is. Sorry for the inconvenience!

**Comment by arousle** [comment]
>You did mention trying absolute paths too, so I guess the issue must be somewhere else if those paths were correct.

Yes, I did try it with absolute paths (`<package path="/home/runner/Steam/steamapps/common/Barotrauma Dedicated Server/Daedalic Entertainment GmbH/BarotraumaWorkshopMods/Installed/2962604067/filelist.xml" />` in the case of my example) as well, with the same result of the content packages not being loaded on server restart.

> We have [a page on the wiki describing how to enable mods on the dedicated server](https://barotraumagame.com/wiki/Enabling_Mods_on_a_Dedicated_Server) as well, please take a look in case you haven't already!

Yes, I've read that as well.

> We also [got another report](https://github.com/Regalis11/Barotrauma/issues/11699) recently about similar difficulties with enabling mods on a Linux dedicated server, although I'm not sure if that's relevant here.

Yeah, I saw that, but found the issue difficult to follow. I don't think it's relevant.

> I'm afraid I can't say for sure where the issue might be off the top of my head, but I hope some of those links can give you some more clues on diagnosing where the problem is. Sorry for the inconvenience!

Thanks. I suppose that as a feature request, would it be possible to add some logging output to the dedicated server launch that prints all content packages being loaded, and also any errors if it fails to load any of the content packages specified in `config_player.xml`? It would make debugging this on the user side a lot easier in the future.

**Comment by nORbDragon** [comment]
First, try adding `enabled="true"` to your packages. Something like that:
```<package path="Daedalic Entertainment GmbH/Barotrauma/WorkshopMods/Installed/2962604067/filelist.xml" enabled="true" />```

If this does not help, then I'll tell you how I set up my game server on linux (CentOS) through `linuxgsm`.

I will not describe the process of installing the game server itself. I will only describe how I installed mods to make them work in a multiplayer game. It's possible that I did it wrong. But at least it works.

1. Copied the mod files to `<linuxgsm folder>/serverfiles/LocalMods/`.
2. Edited `<linuxgsm folder>/serverfiles/config_player.xml`, added strings in `<regularpackages>` block. Only after I added `enabled="true"` did the server start using the mods specified in the file.

Example from my `config_player.xml`:
```  <contentpackages>
    <!--Vanilla-->
    <corepackage
      path="Content/ContentPackages/Vanilla.xml" />
    <regularpackages>
      <package path="LocalMods/2683570256/filelist.xml" enabled="true" />
      <package path="LocalMods/2783973083/filelist.xml" enabled="true" />
      <package path="LocalMods/2893794930/filelist.xml" enabled="true" />
      <package path="LocalMods/2943208698/filelist.xml" enabled="true" />
      <package path="LocalMods/2799051561/filelist.xml" enabled="true" />
      <package path="LocalMods/2101436058/filelist.xml" enabled="true" />
      <package path="LocalMods/2860199840/filelist.xml" enabled="true" />
      <package path="LocalMods/NHC_RS-001/filelist.xml" enabled="true" />
      <package path="LocalMods/Borealis/filelist.xml" enabled="true" />
    </regularpackages>
  </contentpackages>
```

My version requires manually updating the mod files. In my case, mods are not updated automatically through Steam. But for me this is not a problem. The main thing is that the server works and uses the mods I need. If necessary, I update the mods manually.

**Comment by Regalis11** [comment]
@nORbDragon 

> First, try adding enabled="true" to your packages. Something like that:

This does nothing: the only thing the game reads from the package elements is the path (and everything there is already enabled).

**Comment by nORbDragon** [comment]
> @nORbDragon
> 
> > First, try adding enabled="true" to your packages. Something like that:
> 
> This does nothing: the only thing the game reads from the package elements is the path (and everything there is already enabled).

Well, as the developer of this game, you know better.

But it took me several hours to sort through different options for settings, until I came across a mention of this parameter on the Internet. And the server started using the mods.

But again, I could be wrong. However, I suggest that the author of the bug report try adding `enabled="true"` to packages. Suddenly it will help?

**Comment by arousle** [comment]
> ```
>     
>     <corepackage
>       path="Content/ContentPackages/Vanilla.xml" />
>     <regularpackages>
>       <package path="LocalMods/2683570256/filelist.xml" enabled="true" />
>       <package path="LocalMods/2783973083/filelist.xml" enabled="true" />
>       <package path="LocalMods/2893794930/filelist.xml" enabled="true" />
>       <package path="LocalMods/2943208698/filelist.xml" enabled="true" />
>       <package path="LocalMods/2799051561/filelist.xml" enabled="true" />
>       <package path="LocalMods/2101436058/filelist.xml" enabled="true" />
>       <package path="LocalMods/2860199840/filelist.xml" enabled="true" />
>       <package path="LocalMods/NHC_RS-001/filelist.xml" enabled="true" />
>       <package path="LocalMods/Borealis/filelist.xml" enabled="true" />
>     </regularpackages>
>   </contentpackages>
> ```

I see that you're using the `LocalMods` directory — and as a relative path to the game directory, too, which contradicts @Regalis11's earlier statement about requiring absolute paths?

I would expect there to be special handling for local mods, since the game does need to disambiguate between local and subscribed mods somehow(?) — are players able to subscribe to the Steam Workshop mods for the local mods you have enabled, or do they have to download them from the server?

**Comment by Regalis11** [comment]
> and as a relative path to the game directory, too, which contradicts @Regalis11's earlier statement about requiring absolute paths?

The paths should be absolute **unless** the file is inside the game's installation folder, which is the case here. Or in other words, the paths are always relative to the installation folder, unless you specify the full path.

> since the game does need to disambiguate between local and subscribed mods somehow(?) — are players able to subscribe to the Steam Workshop mods for the local mods you have enabled, or do they have to download them from the server?

They should be able to - the Steam IDs of the mods are included in the content package file.

**Comment by nORbDragon** [comment]
Yep, as **Regalis11** above wrote, I store mods in a folder that is inside the game's installation folder. That's why I use the relative path to mods.

About disambiguate between local and subscribed mods: I will write how it works (?) specifically in my case (server installed by linuxgsm, not docker). 

Maybe I'm wrong, but me and my friends, with whom I play on my server, had this impression after experimenting with mods and updates for them.

1. Since I installed "Steam" mods in the "LocalMod" folder, there is no automatic update of the version of these mods. It is possible that there is some way to give a command to Steam on the server that I need to update something somewhere, but I did not look for it. I just take the latest versions of the mods I need from my computer and manually overwrite them on the server in the "LocalMod" folder.
2. When a player connects to my server, the game client automatically downloads all the necessary mods **from the server** to the player's computer. Moreover, if the player already has this mod, but the version of the mod is different, the mod with the "necessary" version will be taken **from the server**, not Steam. And will be used to play on this particular server that version of the mod that is installed on the server.

**Comment by arousle** [comment]
I managed to get it working with:

```xml
<regularpackages>
  <package path="LocalMods/2962604067/filelist.xml" />
</regularpackages>
```

, where `LocalMods` is a subdirectory of the directory that the dedicated server executable is in. 

I'm still not sure why this works while the other package paths I've tried did not, but I suppose my original issue is resolved as I can live with having to use this specific package path definition.

@Regalis11: Would you like me to create a separate feature request for adding the additional log statements I mentioned in [my earlier comment](https://github.com/Regalis11/Barotrauma/issues/11855#issuecomment-1510380353)?

>would it be possible to add some logging output to the dedicated server launch that prints all content packages being loaded, and also any errors if it fails to load any of the content packages specified in config_player.xml? It would make debugging this on the user side a lot easier in the future.

**Comment by YanWenKun** [comment]
Maybe it's relevant to #10112 . A [fix](https://github.com/YanWenKun/barotrauma-server-docker/pull/3/files) is make a nesting symlink:

```
mkdir -p "${GAMEDIR}/Daedalic Entertainment GmbH/Barotrauma/Daedalic Entertainment GmbH"

ln -s \
    "${GAMEDIR}/Daedalic Entertainment GmbH/Barotrauma" \
    "${GAMEDIR}/Daedalic Entertainment GmbH/Barotrauma/Daedalic Entertainment GmbH/Barotrauma"
```

But as @nORbDragon mentioned above, instead of using Steam Workshop mechanism, using `LocalMods` always works fine...

**Comment by Regalis11** [comment]
Seems to be a duplicate of https://github.com/Regalis11/Barotrauma/issues/10112
